text/0000-cargo-schema-version.md

```diff
@@ -0,0 +1,255 @@
+`Cargo.toml` files.  The schema version introduced in this RFC allows Cargo to
+take that version into account in dependency resolution.
+
+This change will also allow for future experiments with new "unstable" Cargo
+features not yet ready for use in stable crates, without allowing those
+features to leak into stable crates or have their behavior set in stone.  Any
+future RFC introducing such behavior should consider whether crates.io should
+prohibit the use of such features entirely, or keep them separated and tagged
+appropriately in indexes.
+
+# Detailed design
+[design]: #detailed-design
+
+Introduce a new Cargo schema version, initially `1.0.0`.  Cargo will increase
+the minor version when introducing any crate-visible change to Cargo behavior
+that old versions of Cargo must not ignore.  (For instance, this would include
+many new `Cargo.toml` stanzas, new environment variables provided to build
+scripts, or assumption of some configuration through convention.)  Cargo would
+increase the major version if it stops handling (or handles incompatibly) some
+behavior that old Cargo handles.  (A major version increase seems unlikely to
+ever happen.)  The patch version will always remain `0`.
+
+Introduce a mechanism for `Cargo.toml` files to specify a minimum major or
+major.minor version for the schema, using the semantics of the `^` operator
+from Cargo semver dependencies.
+
+Versions of Cargo predating the introduction of schema versions must not
+silently ignore the schema requirement; thus, we will use a format change that
+older Cargo will not understand.  `package.name` is currently a mandatory
+field; if missing, old Cargo will stop and reject the package.  So, we will
+move `package.name` (and the other contents of `package`) underneath a new
+`package.major` or `package.major.minor` key.
+
+Packages compatible with schema `0.0` of Cargo (the last Cargo version that
+doesn't support a schema version number) will continue to write `[package]` as
+they do today.  Packages that require schema `1.0` or newer will write:
+
+```toml
+[package.1]
+name = "crate-name"
+```
+
+And packages that require schema `1.5` or newer will write:
+
+```toml
+[package.1.5]
+name = "crate-name"
+```
+
+Semantically, `package` can contain either a single numeric key or the key
+"name".  If it has a single numeric key, use that as the minimum schema major
+version; that key can either contain a single numeric key or the key "name".
+If it has a single numeric key, use that as the minimum schema minor version;
+that key must contain the key "name".
+
+Formalizing this schema as a grammar (for clarity expressed over the parsed and
+normalized hierarchical structure of TOML, rather than the raw text), we have:
+
+```ebnf
+<pkg-meta> ::= { name = ..., version = ..., ... }
+<package> ::= <pkg-meta>
+                    | { <number> = <pkg-meta> }
+                    | { <number> = { <number> = <pkg-meta> } }
+<Cargofile> ::= { package = <package>, ... }
```

Comment:
Rather than attempt to do that in the grammar, I'd rather just add a note in this section mentioning the historical section name and that schema versioning requires `package`.